### PR TITLE
[RELNOTES] Refactor RNNs to rely on atomic cells.

### PR DESCRIPTION
**Pull request by fchollet**
This enables two things:

- instantiating a RNN from a cell instance.
- instantiating an efficient stacked RNN from a list of cells.

Additionally, this removes `implementation=0` mode, as the new design forces RNN logic to be confined to a step function.

In the near future, we should enable CUDNN LSTM and GRU as well.

### CODE REVIEW

**Comment by fchollet** [comment]
@andhus please check it out.

**Comment by fchollet** [comment]
@souptc there seems to be non-obvious CNTK issues going on: https://travis-ci.org/fchollet/keras/jobs/277928028

All tests are passing with TensorFlow and Theano.

**Comment by souptc** [comment]
Let me take a look....

**Comment by souptc** [comment]
@fchollet , there are two issues:
1. when generate dropout mask in the cell:
    `ones = K.ones_like(inputs[:, 0, :]) `
    but previously we are using:
    `ones = K.ones_like(K.reshape(inputs[:, 0, :], (-1, 1))) `
    The reason is cntk's slice op will keep the second dimension as 1 by default, so we need the reshape step to correct the drop out mask's shape.

2. This one is a little tricky. When run K.rnn with unroll is False, CNTK need to convert the timestep axis to a "sequence" axis (which is a cntk concept). To do the element times between input and drop out mask, we need broad cast the dropout mask to the "sequence axis" by:
`C.sequence.broadcast_as(dropout_mask, inputs)`
Previously I handle it in K.rnn, because the dropout_mask is passed in by constant. But now the dropout_mask has been moved to cell, we can't process it in cntk_backend now. Is there any interface that we could pre-process the dropout_mask before start rnn?

**Comment by fchollet** [comment]
Thanks for looking into it! I don't understand the second point:

> When run K.rnn with unroll is False, CNTK need to convert the timestep axis to a "sequence" axis (which is a cntk concept). To do the element times between input and drop out mask, we need broad cast the dropout mask to the "sequence axis"

The dropout masks have no notion of sequence, they have shape `(batch_size, input_dim)` and `(batch_size, output_dim)` respectively. Because they are time-independent, they were moved inside the step function. Similarly the contents of the step function have no notion of sequence.

**Comment by souptc** [comment]
right, here keras expect the input of the cell is also time-independent, with shape (batch_size, input_dim), and RNN run those cells with a while loop. But the tricky thing is **CNTK doesn't have a symbolic loop**, instead it rely on a hidden "sequence" to do the recurrence. So the input to cell here have a hidden "sequence" in cntk while the dropout mask doesn't have, that's why I need the `C.sequence.broadcast_as(dropout_mask, inputs)`. 

Another choice is override the cntk element-wise op for keras case, to auto-broadcast it. But that need change in CNTK python package.

**Comment by fchollet** [comment]
Ok, I see. I think we should disable RNN dropout when using unrolled RNNs in CNTK (if unrolled and dropout, print a warning and set dropout to 0), until there is a solution to have the step function be truly time-independent.

People writing their own custom RNNs (which is the main purpose of this refactor) would likely run into the same issue if the step function doesn't behave in a truly time-independent way.

**Comment by souptc** [comment]
Should it be "not-unrolled RNN"? this issue only happened when unrolled is False.

**Comment by fchollet** [comment]
Sorry yes, that's what I meant.

**Comment by souptc** [comment]
:) yes, we can disable it for cntk now. Actually internally we have a cntk related project which will have symbolic loop, we will take a look and try to enable this as soon as we can.

**Comment by andhus** [comment]
@fchollet great that this was prioritised, I'll look through it and will adapt https://github.com/fchollet/keras/issues/7633 accordingly!
